Support for Parallel triggered builds

tdctaz opened the pull request:
This adds functionality for triggering schedulers multiple times, for example when we want to split a task into multiple partitions for parallelization purposes.

- Adds a build step for triggering multiple copies of the same scheduler with different properties
- Extends `FindPreviousSuccessfulBuild` and `CheckArtifactExists` to take any configured merge function into account (This is a breaking change). If no merge function has been configured then it works as before, so the extra overhead for getting the data for the merge function is only added when it is there.
- Finally adds the option for a triggered build to download artifacts from the parent build that triggered it

unity-christianb committed:
Initial version

unity-christianb committed:
Cleaned up code

MindWrapper committed:
added possibility to download artifact from a partn build

unity-christianb committed:
Make previous build take mergeFunction into account

unity-christianb committed:
Optimized find previous build

unity-christianb committed:
Remove limit from build requests

determining the amount of previous build request we would need is not easy

unity-christianb committed:
Added tests for partition trigger

unity-christianb committed:
Finished support for find previous successful build

tdctaz commented on master/buildbot/steps/artifact.py:
```diff
@@ -35,7 +37,64 @@ def forceRebuild(build):
+
+class FindPreviousSuccessBuildMixin():
+    @defer.inlineCallbacks
+    def _determinPreviousBuild(self, master, build, build_sourcestamps):
```
Spelling `_determinePreviousBuild`

unity-christianb committed:
Ignore files generated by tests

tdctaz commented on master/buildbot/steps/artifact.py:
```diff
@@ -35,7 +37,64 @@ def forceRebuild(build):
+    # A more light weight function for making a build request, it uses as much of the information we already have to
+    # make a build request - so we only need to fetch the properties
+    @defer.inlineCallbacks
+    def _getBuildRequest(self, master, brdict, sources):
```
`n+1` selects, I'll try to batch select them outside the loop and then split the results inside here

tdctaz commented on master/buildbot/steps/artifact.py:
```diff
@@ -366,7 +423,9 @@ def start(self):
         triggeredbybrid = self.build.requests[0].id
         br = yield self.master.db.buildrequests.getBuildRequestTriggered(triggeredbybrid, self.artifactBuilderName)
-
+        if br is None:
```
Make new subclass instead of this logic, perhaps `DownloadArtifactFromTriggered`?

unity-christianb committed:
Fix: N+1 Select of properties and buildsets

unity-christianb committed:
Fix: Spelling error

unity-christianb committed:
Refactor: Split DownloadArtifact into DownloadArtifactFromParent and DownloadArtifact

MindWrapper committed:
added method to get all builder triggered by a given build-request id

mariangemarcano commented on master/buildbot/test/unit/test_steps_partition_trigger.py:
```diff
@@ -0,0 +1,202 @@
+    def test_runStep_whenPartitionFunctionReturnsNoBuilds_thenNoSchedulesIsTriggered(self):
+        def yieldNoPartitions(buildStep, scheduler):
+            if False:
+                yield {'partition-index': -1}
```
if we dont have a partition then katana would just run the scheduler as usual

unity-christianb committed:
Change: If no parition then normal build is triggered

- also fixed the status text was not compared, when just assigned a single string and not an array.

MindWrapper committed:
added test for buildrequests.getBuildRequestsTriggeredBy

MindWrapper committed:
test cleanup

MindWrapper committed:
testing download artifact from children

MindWrapper committed:
some code cleanup

MindWrapper committed:
more cleanup.  added step description adn stdio

MindWrapper committed:
changed Download artifacts from children so that it inherits from the LoggedBuildStep

MindWrapper committed:
extended test case to check that 'artifactbrid' is used when specified

MindWrapper committed:
made local dir optional. fixed some issues in tests

MindWrapper committed:
made mkdir platform independent

unity-christianb committed:
Add option for disabling the Unmergable functionality

tdctaz commented on master/buildbot/steps/artifact.py:
```diff
@@ -379,6 +448,108 @@ def start(self):
+        self.master = None
+        self.usePowerShell = usePowerShell
+        self.kwargs = kwargs
+        self.name = name
```
`name`, `description` and `descriptionDone` should be automatic generated or just constants on the class

tdctaz commented on master/buildbot/steps/artifact.py:
```diff
@@ -379,6 +448,108 @@ def start(self):
+        br = yield self.master.db.buildrequests.getBuildRequestById(id)
+        defer.returnValue(br)
+
+class DownloadArtifactFromChilden(LoggingBuildStep, CompositeStepMixin):
```
Spelling error, lets call it: `DownloadArtifactsFromChildren` since it download multiple artifacts and fix childen -> children

tdctaz commented on master/buildbot/steps/artifact.py:
```diff
@@ -379,6 +448,108 @@ def start(self):
+        self.workdir = 'build'
+        self.artifactBuilderName = artifactBuilderName
+        self.artifact = artifact
+        self.projectPrefix = projectPrefix
```
`projectPrefix` is a Unity concept, dont think it should be a part of the download process

tdctaz commented on master/buildbot/steps/artifact.py:
```diff
@@ -379,6 +448,108 @@ def start(self):
+                 artifact=None, artifactDirectory=None, artifactDestination=None,
+                 artifactServer=None, artifactServerDir=None, artifactServerPort=None,
+                 usePowerShell=True, baseLocalDir=None, **kwargs):
+        self.workdir = 'build'
```
what is the `workdir` it seams to be a unity specific thing

tdctaz commented on master/buildbot/steps/artifact.py:
```diff
@@ -379,6 +448,108 @@ def start(self):
+            self.master = self.build.builder.botmaster.parent
+
+        brid = self.build.requests[0].id
+        triggeredBuilderName = self.projectPrefix + self.target_config
```
Should this not just be using `artifactBuilderName` instead?

tdctaz replied:
lets remove this line and change the line below to `partitionRequests = yield self.master.db.buildrequests.getBuildRequestsTriggeredBy(brid, self.artifactBuilderName)`

tdctaz commented on master/buildbot/steps/artifact.py:
```diff
@@ -379,6 +448,108 @@ def start(self):
+
+        for id in buildRequetsIdsWithArtifacts:
+            buildRequest = yield self.master.db.buildrequests.getBuildRequestById(id)
+            triggeredBuilderName = self.projectPrefix + self.target_config
```
Same issue should just be using `self.artifactBuilderName`

tdctaz commented on master/buildbot/steps/artifact.py:
```diff
@@ -379,6 +448,108 @@ def start(self):
+
+            artifactPath += "/%s" % self.artifactDirectory
+            remotelocation = getRemoteLocation(self.artifactServer, self.artifactServerDir, artifactPath, "")
+            localdir = self.baseLocalDir + '/' + str(id)
```
if `baseLocalDir` is `None` to should copy to the root - need check for `None`

tdctaz commented on master/buildbot/steps/artifact.py:
```diff
@@ -379,6 +448,108 @@ def start(self):
+                 usePowerShell=True, baseLocalDir=None, **kwargs):
+        self.workdir = 'build'
+        self.artifactBuilderName = artifactBuilderName
+        self.artifact = artifact
```
Is this used? - could we make it possible to download one specific artifact from all children if it is specified?

tdctaz commented on master/buildbot/steps/artifact.py:
```diff
@@ -379,6 +448,108 @@ def start(self):
+        defer.returnValue(br)
+
+class DownloadArtifactFromChilden(LoggingBuildStep, CompositeStepMixin):
+    def __init__(self, name=None, projectPrefix=None, targetConfig=None, artifactBuilderName=None,
```
Lets not have default `None` for all the required fields

tdctaz commented on master/buildbot/steps/artifact.py:
```diff
@@ -379,6 +448,108 @@ def start(self):
+            command = [mkDir(self), '-p', '\''+ localdir + '\'']
+            yield self._docmd(command)
+
+            rsync = rsyncWithRetry(self, remotelocation, localdir, self.artifactServerPort)
```
How are errors handled? - in my test I tried to download a non-existing artifact folder and got the following error:

`rsync: change_dir "/home/vagrant/artifacts/Simple_Test_661_02_07_2017_21_28_52_+0000/foobar" failed: No such file or directory (2)`

but the step still said `Success`

tdctaz commented on master/buildbot/steps/artifact.py:
```diff
@@ -379,6 +448,108 @@ def start(self):
+            remotelocation = getRemoteLocation(self.artifactServer, self.artifactServerDir, artifactPath, "")
+            localdir = self.baseLocalDir + '/' + str(id)
+
+            command = [mkDir(self), '-p', '\''+ localdir + '\'']
```
not sure if it is this command but the destination directory is really strange with duplicate of each directory, here is the `ls -la` of the build directory:

```drwx------ 8 vagrant vagrant 4096 Jul  3 14:29 ./
drwx------ 3 vagrant vagrant 4096 Jul  3 14:29 ../
drwxrwxr-x 2 vagrant vagrant 4096 Jul  3 14:29 845/
drwx------ 2 vagrant vagrant 4096 Jul  3 14:29 '845'/
drwxrwxr-x 2 vagrant vagrant 4096 Jul  3 14:29 846/
drwx------ 2 vagrant vagrant 4096 Jul  3 14:29 '846'/
drwxrwxr-x 2 vagrant vagrant 4096 Jul  3 14:29 848/
drwx------ 2 vagrant vagrant 4096 Jul  3 14:29 '848'/
```

tdctaz replied:
only the folder without `'` has the downloaded artifact within them

MindWrapper committed:
fixing various issues found during code review #1

MindWrapper committed:
more cleanup

MindWrapper committed:
do not download artifacts from merged partitions twice

MindWrapper committed:
added comment what a hell test is doing

MindWrapper committed:
some refactoring

MindWrapper committed:
setup command behaviors to return rc=0

MindWrapper committed:
removed unreacheble code

MindWrapper committed:
fixed compilation issue

unity-christianb committed:
Made artifact port optional

unity-christianb committed:
Added more nice descriptions while downloading

unity-christianb committed:
Remove DownloadFromParent

it will not be needed in the near future

unity-christianb committed:
Fix: Broken test / broken code

mariangemarcano commented on master/buildbot/db/buildrequests.py:
```diff
@@ -495,6 +495,40 @@ def thd(conn):
+                .where(buildrequests_tbl.c.results == 0) \
+                .where(buildrequests_tbl.c.buildername == buildername) \
+                .where(buildrequests_tbl.c.artifactbrid == None) \
+                .select_from(buildrequests_tbl.outerjoin(claims_tbl, (buildrequests_tbl.c.id == claims_tbl.c.brid))) \
```
@tdctaz I think it is ok to skip this join (as we are doing in the original query). Did you  add it for some reason?

tdctaz replied:
@mariangemarcano To be honest I can't remember - I think I just looked at the requirements for `_brdictFromRow` and selected what it needed. - Will remove and retest.

mariangemarcano commented on master/buildbot/db/buildrequests.py:
```diff
@@ -755,6 +789,72 @@ def thd(conn):
+
+        return self.db.pool.do(thd)
+
+    def getTriggeredById(self, id):
```
@tdctaz it looks that we are not longer using this one, could we remore it?

tdctaz replied:
Yes 👍

mariangemarcano commented on master/buildbot/steps/artifact.py:
```diff
@@ -298,6 +362,13 @@ def rsyncWithRetry(step, origin, destination, port=None):
 
+def mkDir(step):
+    if _isWindowsSlave(step):
+        return r'C:\cygwin64\bin\mkdir.exe'
```
the previous version we only did mkdir, why do we need to change it?

MindWrapper replied:
@mariangemarcano if I don't do it - it will invoke a Windows version of mkdir, which does not support nested dirs creation. Maybe it would work ok a katana build agent, but it will definitely fail on my local build agent.

mariangemarcano replied:
@MindWrapper I think we better remove it since we dont require it in our production setup

campos-ddc replied:
Even if our production setup we have to call mkdir from cygwin explicitly, since the `mkdir` function from cmd takes precedence over `mkdir.exe` in PATH

mariangemarcano replied:
@MindWrapper I guess we need to check with @opdude / @jwebbintx if there are ok with this location on the agents (I'm not sure if the location is consistent among windows setup) or what works best for our setup (make it configurable?)

opdude replied:
I certainly think it's a bad idea to put this in like this, this is 100% unity specific and even then it's 100% agent specific, please use mkdir.exe and have people put it in their PATHs instead

MindWrapper replied:
@mariangemarcano @opdude @campos-ddc 
The problem with mkdir is that it is one of the commands included into the command.com. it means that it does not matter what are you going to add into the PATH it will be executed first
`createDir` in global.config does it the same way. I'll try to invent something wich does not require hard-coding it in this part of the code

MindWrapper replied:
Fixed it in 0fdcc4e and 982af46

mariangemarcano commented on master/buildbot/steps/artifact.py:
```diff
@@ -402,6 +472,113 @@ def start(self):
+
+        self._setUpLogs()
+
+        partitionRequests = yield self.master.db.buildrequests.getBuildRequestsTriggeredBy(self.build.requests[0].id, self.artifactBuilderName)
```
if the partitions are never merge by different builders then is ok, otherwise we are going to hit problems when 2 different builders had a partition merged and try to download.

tdctaz replied:
@mariangemarcano Not sure what you mean? - so are we talking about the child build requests being merged? - the next function `_getBuildRequestIdsWithArtifacts` should sort it out.

If the build request has an `artifactbrid` then we will use that as the real build request otherwise we will use the `brid`. Should the build request be merged at the point of creating the partitions the id's will be duplicated and the duplicates will get removed.

We did some test in the loadtests setup. We can take a look at it tomorrow if you want just so we are sure we haven't misunderstood something.

mariangemarcano replied:
ok @tdctaz  that should work it out ignore this comment 👍

mariangemarcano commented:
@tdctaz added a few comments and included @campos-ddc to help checking the PR

unity-christianb committed:
Remove unused method

unity-christianb committed:
Remove claims from the build requests

mariangemarcano commented on master/buildbot/steps/artifact.py:
```diff
@@ -402,6 +472,113 @@ def start(self):
+
+        self._setUpLogs()
+
+        partitionRequests = yield self.master.db.buildrequests.getBuildRequestsTriggeredBy(self.build.requests[0].id, self.artifactBuilderName)
```
@tdctaz ok just to be sure let's check again tomorrow so we make sure that the conditions where the triggered builds getting merged is sorted out.

unity-christianb committed:
Refactor: lower case `workdir` to match `RemoteShellCommand`

unity-christianb committed:
Fix: workDir casing in tests

MindWrapper committed:
removed hardcoded path to cygwin

MindWrapper committed:
fixed broken test

MindWrapper committed:
beautified path transforming

MindWrapper committed:
moved import to the import section

MindWrapper committed:
made interfact compatible with LogFile

MindWrapper committed:
set content type for fake object (same as katana does)

MindWrapper committed:
applied the change for artifacts path calulation

MindWrapper committed:
make 'resume' smart again

MindWrapper committed:
fixed broken tests

tdctaz commented on master/buildbot/test/unit/test_step_artifact.py:
```diff
@@ -314,3 +313,121 @@ def test_download_artifact_reusing_build(self):
+                artifactDestination='./base/local',
+        ), [br2, br3, br4, br666])
+
+        expectedRemote1 = '\'usr@srv.com:/artifacts/B_2_01_01_1970_00_00_00_+0000/\''
```
This cannot be correct? the `artifactDirectory` is defined to `mydir` so the expected remote should be `usr@srv.com:/artifacts/B_2_01_01_1970_00_00_00_+0000/mydir` right?

tdctaz commented on master/buildbot/test/unit/test_step_artifact.py:
```diff
@@ -314,3 +313,121 @@ def test_download_artifact_reusing_build(self):
+                artifactDestination='./base/local'
+        ), [br2], winslave=True)
+
+        expectedRemote = '\'usr@srv.com:/artifacts/B_2_01_01_1970_00_00_00_+0000/\''
```
This cannot be correct? the `artifactDirectory` is defined to `mydir` so the expected remote should be `usr@srv.com:/artifacts/B_2_01_01_1970_00_00_00_+0000/mydir` right?

unity-christianb committed:
Add logging in an attempt to find bug on staging

unity-christianb committed:
Remove sql debug messages

did keep the merge message and fixed the tests

MindWrapper committed:
filling artifact map

unity-christianb committed:
Revert bad modification of tests and fixed artifact.py

mariangemarcano commented on master/buildbot/steps/artifact.py:
```diff
@@ -39,7 +45,76 @@ def forceRebuild(build):
+            defer.returnValue((PreviousBuildStatus.Unmergeable, None))
+
+        elif mergeRequestFn == True or mergeRequestFn == False: # Default case, with no merge function assigned
+            log.msg("[brid: %d] is searcing for previous successful build without merge function" % build.requests[0].id)
```
typo searcing => searching (the typo needs to be fix in several places)

mariangemarcano commented on master/buildbot/steps/artifact.py:
```diff
@@ -39,7 +45,76 @@ def forceRebuild(build):
+
+        else: # Custom merge function assigned
+            log.msg("[brid: %d] is searcing for previous successful build with merge function" % build.requests[0].id)
+            prevBuildRequests = yield master.db.buildrequests \
```
I guess prevBuildRequests can be put outside the if/else, but is fine leaving it as it is, I didn't catch it in the initial review.

tdctaz replied:
No sure why it should be move outside the if/else it would only add overhead by running the query event if we are no inside the case

mariangemarcano replied:
@tdctaz we are not breaking execution(returning) after defer.returnValue, it's fine to leave the code as it is

mariangemarcano commented on master/buildbot/steps/artifact.py:
```diff
@@ -402,6 +482,133 @@ def start(self):
+        for brid in buildRequetsIdsWithArtifacts:
+            buildRequest = yield self.master.db.buildrequests.getBuildRequestById(brid)
+
+            localdir = self._getLocalDir(brid)
```
DownloadArtifactsFromChildren has the local directory fix with a specific structure, I guess it is ok for this use case.

tdctaz replied:
Yes it adds the build request id to the path, in order to separate the downloaded artifacts. For example if we download artifacts from build; `1`, `3`, `7`, we want the artifacts from those 3 build to be in separate locations, eg. subfolder `1`, `3` and `7` so they dont overwrite each others artifacts.

mariangemarcano reviewed (approved):
Just minor comments

MindWrapper committed:
Applying fix from this PR:
Fix a rare case where a buildtep might fail from a GeneratorExit exception (pull request 1063
(https://github.com/buildbot/buildbot/pull/1063)

unity-christianb committed:
Fix: Spelling mistake

unity-christianb committed:
Log attempts

MindWrapper committed:
try to do not consume errors

MindWrapper committed:
do not set step name

unity-christianb committed:
Revert "try to do not consume errors"

This reverts commit 8a2a96751a2ab38502ec037f9c2232c7d7973232.

unity-christianb committed:
Lets log when a build request is being reused

MindWrapper committed:
fixed broken tests

mariangemarcano commented on master/buildbot/schedulers/base.py:
```diff
@@ -412,8 +412,8 @@ def addBuildsetForSourceStampSetDetails(self, reason, sourcestamps,
                 if brids and brids.keys() == builderNames:
+                    log.msg("Resuing existing buildrequest %s for trigger by %s" % (brids, triggeredbybrid))
                     defer.returnValue((bsid, brids))
-                    return
```
@tdctaz this changed seems unintentional I think we should undo it

mariangemarcano replied:
@tdctaz as we talked we are going to remove this logic, which can cause issues from line 398-415 and also remove getBuildRequestsTriggeredByScheduler which incorrectly filters by property

unity-christianb committed:
Remove retry logic based on stepname
